### PR TITLE
Data mover micro-service watcher

### LINKED ISSUE #7932: Data mover micro service - dual mode watcher
Add the dual mode watcher according to design #7576 for data mover micro service

### PR DESCRIPTION
**Pull request by Lyndon-Li**
Fix #7932. Data mover micro-service watcher according to design https://github.com/vmware-tanzu/velero/pull/7576

### CODE REVIEW

**Comment by kaovilai** [comment]
I think re-titling `ms` to `microservice` would be less confusing.

**Comment by Lyndon-Li** [comment]
> I think re-titling `ms` to `microservice` would be less confusing.

Done

**Review by blackpiglet — DISMISSED** [review]

**Review by sseago — DISMISSED** [review]

File: pkg/datapath/micro_service_watcher.go [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+		},
+	)
+
+	if err != nil {
+		return errors.Wrap(err, "error registering pod handler")
+	}
+
+	defer func() {
+		if !succeeded {
+			if err := podInformer.RemoveEventHandler(podHandler); err != nil {
```
Comment (shubham-pampattiwar) [inline pkg/datapath/micro_service_watcher.go:175]: Shouldn't we use `ms.Close()` here ? Are we not missing the `cancel()` call as well ? 
same as line 148 comment earlier. Curious to know if my understanding is incorrect.

Comment (shubham-pampattiwar) [reply pkg/datapath/micro_service_watcher.go:175]: or do we need to combine both the defer calls for podhandler and eventhandler and then call `ms.Close()`

File: pkg/datapath/micro_service_watcher.go [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+		},
+	)
+
+	if err != nil {
+		return errors.Wrap(err, "error registering event handler")
+	}
+
+	defer func() {
+		if !succeeded {
+			if err := eventInformer.RemoveEventHandler(eventHandler); err != nil {
```
[the remark this answers is not in the document]
Comment (Lyndon-Li) [reply pkg/datapath/micro_service_watcher.go:147]: I am not so confident to call `ms.Close()` in the `Init`, as `ms.Close()` is doing more things and may do even more in future, if it is not required, we should not couple them together.
So here I moved `ms.ctx, ms.cancel = context.WithCancel(ctx) ` to the end, actually, this ctx with cancel is used after Init.

File: pkg/datapath/micro_service_watcher.go (hunk 1 above)
Comment (Lyndon-Li) [reply pkg/datapath/micro_service_watcher.go:174]: See https://github.com/vmware-tanzu/velero/pull/7999#discussion_r1690716165

File: pkg/datapath/micro_service_watcher.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Lyndon-Li) [reply pkg/datapath/micro_service_watcher.go:147]: Actually, the caller may also call `ms.Close()` when `Init` returns error. At present, there is no problem.
But here I think we should avoid adding one more coupling if we can, in future, if `ms.Close()` is not suitable to be called when `Init` returns error, we should correct the caller.

**Comment by Lyndon-Li** [comment]
@shubham-pampattiwar @sseago @blackpiglet I changed the code slightly per discussion https://github.com/vmware-tanzu/velero/pull/7999#discussion_r1690292825, please have another look. Thanks.

**Review by shubham-pampattiwar — APPROVED** [review]

**Review by sseago — APPROVED** [review]

**Pull request merged**